### PR TITLE
Adding usual Application Menu

### LINKED ISSUE #98: Unable to copy/paste in any field (Mac 10.11.3)
Would like to be able to copy / paste within a password/keyword/ and pin field for the Safe Launcher (essential for strong passwords).

### PR DESCRIPTION
**Pull request by ligthyear**
As described in #98 , adding an application menu to electron fixes the copy-paste problem on mac. This PR adds that menu.

fixes #98 .

---
This change is [Reviewable](https://reviewable.io/reviews/maidsafe/safe_launcher/120)

### CODE REVIEW

**Comment by maidsafe-highfive** [comment]
Thanks for the pull request, and welcome! The MaidSafe team is excited to review your changes, and you should hear from @krishnaIndia (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. The way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [CONTRIBUTOR.md](../blob/master/CONTRIBUTOR) for more information.

**Comment by krishnaIndia** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[app/app.js, line 13 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120#-KD4ypI2A9Oxmgoom8Lu:-KD4ypI2A9Oxmgoom8Lv:-1880508819) ([raw file](https://github.com/maidsafe/safe_launcher/blob/7c2b94184e8d82d921f877f92f24a533c8aca71f/app/app.js#L13)):*
Setting the menu is managed from [background.js](https://github.com/maidsafe/safe_launcher/blob/master/app/background.js#L20).  If we can move this code to the same place then managing it can be easier.

---

*[app/app.js, line 128 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120#-KD4zUYFKuBiq9LqZ1qU:-KD4zUYFKuBiq9LqZ1qV:-1643310854) ([raw file](https://github.com/maidsafe/safe_launcher/blob/7c2b94184e8d82d921f877f92f24a533c8aca71f/app/app.js#L128)):*
Can we have a simpler template for menu with just `cut`, `copy` and `paste`? I don't  see a point why we would need Undo, Redo etc.
I guess just `cut`, `copy` and `paste` would be good. Please do let me know if I am wrong.

---

*[app/app.js, line 277 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120#-KD5-RU8AJeGaCO1hlh8:-KD5-RU9FsZdL3mEMAgs:40085310) ([raw file](https://github.com/maidsafe/safe_launcher/blob/7c2b94184e8d82d921f877f92f24a533c8aca71f/app/app.js#L277)):*

As per the UI design the menu bar shouldn't be visible.  Can we hide the menu bar using `mainWindow.setMenuBarVisibility(false);`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120)*

**Comment by krishnaIndia** [comment]
Thanks @ligthyear for the fix. Once the points raised are addressed, we can test and get the code merged.

**Comment by ligthyear** [comment]
Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions.

---

*[app/app.js, line 13 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120#-KD4ypI2A9Oxmgoom8Lu:-KD5Ar8DOpEzB-QkIHh_:900023263) ([raw file](https://github.com/maidsafe/safe_launcher/blob/7c2b94184e8d82d921f877f92f24a533c8aca71f/app/app.js#L13)):*
I was wondering where that came from. Of course I can move the stuff there.

---

*[app/app.js, line 128 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120#-KD4zUYFKuBiq9LqZ1qU:-KD5B6ZkeqYM4xjr8CeY:-1540207676) ([raw file](https://github.com/maidsafe/safe_launcher/blob/7c2b94184e8d82d921f877f92f24a533c8aca71f/app/app.js#L128)):*
Yeah, sure. I just went for the example that Electron provided. Will slim it down.

---

*[app/app.js, line 277 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120#-KD5-RU8AJeGaCO1hlh8:-KD5BKCmufQsMFJY9L46:210328092) ([raw file](https://github.com/maidsafe/safe_launcher/blob/7c2b94184e8d82d921f877f92f24a533c8aca71f/app/app.js#L277)):*
I will try it and see if the fix still works. If it does I am happy to do that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120)*

**Comment by ligthyear** [comment]
@krishnaIndia I merge the dev helper menu with this new one. All details in [255a849](https://github.com/maidsafe/safe_launcher/pull/120/commits/255a849b8d91afe5395e7b00d89875f70345b8b0). I also added the requested hiding of the menu but for me here on Mac that has no effect.

**Comment by krishnaIndia** [comment]
@ligthyear I will also check the same on a Mac machine. If I find a work around, will update the same here.

**Comment by ligthyear** [comment]
@krishnaIndia anything? not having to copy-paste on mac is a real pain ...

**Comment by ligthyear** [comment]
Review status: 0 of 3 files reviewed at latest revision, 3 unresolved discussions.

---

*[app/app.js, line 13 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120#-KD4ypI2A9Oxmgoom8Lu:-KH4A6XRhWfz4BmLJnQN:-499146009) ([raw file](https://github.com/maidsafe/safe_launcher/blob/7c2b94184e8d82d921f877f92f24a533c8aca71f/app/app.js#L13)):*
Done.

---

*[app/app.js, line 128 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120#-KD4zUYFKuBiq9LqZ1qU:-KH4A3coHblsmPgkIKMs:-499146009) ([raw file](https://github.com/maidsafe/safe_launcher/blob/7c2b94184e8d82d921f877f92f24a533c8aca71f/app/app.js#L128)):*
Done.

---

*[app/app.js, line 277 \[r1\]](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120#-KD5-RU8AJeGaCO1hlh8:-KH4A5VHuXt5pLx5eZjQ:-499146009) ([raw file](https://github.com/maidsafe/safe_launcher/blob/7c2b94184e8d82d921f877f92f24a533c8aca71f/app/app.js#L277)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120)*

**Comment by krishnaIndia** [comment]
@ligthyear, the restriction to copy and  paste in the password fields in OS-X desktop apps seems to be enforced by the OS intentionally. The [stackoverflow comment](http://apple.stackexchange.com/questions/79986/why-cant-i-paste-into-the-password-dialog-when-mounting-an-encrypted-disk-image) also explains about the same. I think any workaround to fix it will be going against the standards that is being enforced. For OS-X we should leave it as is and not try to workaround the restriction enforced, IMO. What do you think? If you agree then we can close this PR.

**Comment by ligthyear** [comment]
No, I **strongly** disagree. The notion that's its somehow "the standard on
OSX" is bollocks. The thread on stack overflow you linked to is 3years old.
I am a full time Mac user, as a freelancer in sensitive areas having a
fully encrypted disk. Aside from the login screen (where there's no way to
copy information from) I can copy-paste into _all password fields_
including all apple products like iTunes and the AppStore. I can assure
you: I am solely using a password manager (keepassX) for years and don't
remember any passwords (40chars are just to long to remember).

I also thought we are past that point of believing that disallowing
copy-paste makes for more security. It has been debunked so often (even
twice in that same link you posted), while shown clear usability and
security problems as people then go for less secure passwords they can
remember and/or type. And as we are asking for three fields, all of which
hide the keys typed, disallowing copy-paste is a nightmare to use. Besides
if that was so good as a security feature shouldn't we enforce it across
all platforms? (The answer of course is: no, because it is not a security
feature.)

To recap: its not standard (quite the opposite) and it's not any more
secure (also quite the opposite).

Random question: is anyone in the team using the launcher on a OSX?

**Comment by ligthyear** [comment]
If it can't be merged easily, I am happy to rebase my changes and force push that again.

**Comment by krishnaIndia** [comment]
> Random question: is anyone in the team using the launcher on a OSX?

Yes, there are people using OSX. @Viv-Rajkumar uses OSX.

Shall we will also wait for his opinion on this?

**Comment by ligthyear** [comment]
> Shall we also wait for his opinion on this?

That's up to you. I was just asking because it's kinda the first thing you notice (and a big pain point) when you open the launcher in OSX. Was just wondering how other OSX users "cope" with it...

**Comment by Viv-Rajkumar** [comment]
While I agree if its just a bug in electron impl we can get this patched, dont think

> Aside from the login screen (where there's no way to copy information from) I can copy-paste into *all password fields* including all apple products like iTunes and the AppStore.

is always the case. I'm using 10.10 myself and also use a password manager and have had a few places where paste in native apps are still disabled. which is also why I use the dock applescript plugin to workaround those cases. For example "System Preferences" and choose unlock and the admin password confirmation dialog has the same Copy / Paste disabled. Same for some options in Disk Utility where you are creating a new local mount with custom credentials.

Back on topic tho, if this is merely just a bug in electron and this patch addresses the issue, don't see any issue with it then.

**Comment by ligthyear** [comment]
I'm a little confused. This has been closed without merge, nor fixing the issue reported in #98, and no comment on why it was closed although @Viv-Rajkumar agreed that this being primarily a bug in electron he doesn't see a reason to not merging it (and even agreeing that not being able to copy-paste in is bad UX).

As said before, if the main reason for that is, that it needs to be rebased, I am happy to do that!

**Comment by krishnaIndia** [comment]
@lightyear, I am really sorry. I should have closed it by accident. This will be merged for sure.
Can you please rebase to the current master branch please

**Comment by ligthyear** [comment]
Rebased, tested (can copy-paste) and pushed – no code changes. @krishnaIndia : it's ready for review/merge.

**Comment by krishnaIndia** [comment]
Reviewed 3 of 4 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/maidsafe/safe_launcher/120)*
